src/qmapshack/poi/CPoiPOI.cpp

```diff
@@ -57,9 +66,128 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
     poi->convertRad2Px(pp);
     p.translate(-pp);
 
+    // calculate maximum viewport
+    qreal x1 = buf.ref1.x() < buf.ref4.x() ? buf.ref1.x() : buf.ref4.x();
+    qreal y1 = buf.ref1.y() > buf.ref2.y() ? buf.ref1.y() : buf.ref2.y();
+
+    qreal x2 = buf.ref2.x() > buf.ref3.x() ? buf.ref2.x() : buf.ref3.x();
+    qreal y2 = buf.ref3.y() < buf.ref4.y() ? buf.ref3.y() : buf.ref4.y();
+
+    if(x1 < -180.0 * DEG_TO_RAD)
+    {
+        x1 = -180 * DEG_TO_RAD;
+    }
+    if(x2 >  180.0 * DEG_TO_RAD)
+    {
+        x2 =  180 * DEG_TO_RAD;
+    }
+
     // draw POI
-    QPointF myPoi(12.329699 * DEG_TO_RAD, 45.437951 * DEG_TO_RAD);
-    poi->convertRad2Px(myPoi);
-    QPixmap icon("://icons/waypoints/32x32/FlagBlue.png");
-    p.drawPixmap(myPoi, icon);
+    mutex.lock(); //Does this have to be so broad?
+    for(QString category : categoryActivated.keys(true))
+    {
+        for(int minLonM10 = qFloor(x1 * RAD_TO_DEG * 10); minLonM10 < qCeil(x2 * RAD_TO_DEG * 10); minLonM10++)
+        {
+            for(int minLatM10 = qFloor(y2 * RAD_TO_DEG * 10); minLatM10 < qCeil(y1 * RAD_TO_DEG * 10); minLatM10++)
+            {
+                if(!loadedPOIs.contains(category) ||
+                   !loadedPOIs[category].contains(minLonM10) ||
+                   !loadedPOIs[category][minLonM10].contains(minLatM10))
+                {
+                    mutex.unlock();
+                    loadPOIsFromFile(category, minLonM10, minLatM10);
+                    mutex.lock();
+                }
+
+                for(poi_t poiToDraw : loadedPOIs[category][minLonM10][minLatM10])
```

Comment:
`const poi_t& poiToDraw`

Reply:
That's not possible since below there is a call to convertRad2Px which changes the values in-place.

Reply:
ok in that case `poi_t& poiToDraw`. The important part is the reference.

Reply:
But if I do the reference it will be changed in the QMap. And if it is drawn the next time, no longer the rad value will be converted to pixel, but the old pixel value will be converted to something else. Or do I get something wrong?

Reply:
Ok, that slipped my attention, loading POIs is accumulative. I wonder if that is a good idea as it will consume more and more memory. On the other hand side it shouldn't be that much.

I would prefer some on-the-fly query however the whole process seems to take quite some time. Haven't placed timers yet to find out what part consumes the most. But we can do the optimization in another ticket.

Reply:
```
    p.translate(-pp);
    p.setBrush(Qt::white);
    p.setPen(Qt::NoPen);
...
...
This is my slightly enhanced version:
...
                if(poi->needsRedraw())
                {
                    mutex.unlock();
                    return;
                }

                for(const poi_t& poiToDraw : loadedPOIs[category][minLonM10][minLatM10])
                {
                    QPointF pt = poiToDraw.coordinates;
                    poi->convertRad2Px(pt);

                    QRectF r1 = icon.rect();
                    r1.moveCenter(pt);

                    p.drawRect(r1.adjusted(-1, -1, 1, 1));
                    p.drawPixmap(r1.topLeft(), icon);
                }

```